src/rgw/rgw_swift_auth.cc

```diff
@@ -10,15 +10,298 @@
+                                   bucket_info, nullptr);
+  if (ret < 0) {
+    throw ret;
+  }
+
+  ldout(s->cct, 20) << "temp url user (bucket owner): " << bucket_info.owner
+                    << dendl;
+
+  if (rgw_get_user_info_by_uid(store, bucket_info.owner, owner_info) < 0) {
+    throw -EPERM;
+  }
+}
+
+bool RGWTempURLAuthEngine::is_expired(const std::string& expires) const
+{
+  string err;
+  const utime_t now = ceph_clock_now(g_ceph_context);
+  const uint64_t expiration = (uint64_t)strict_strtoll(expires.c_str(),
+                                                       10, &err);
+  if (!err.empty()) {
+    dout(5) << "failed to parse temp_url_expires: " << err << dendl;
+    return true;
+  }
+
+  if (expiration <= (uint64_t)now.sec()) {
+    dout(5) << "temp url expired: " << expiration << " <= " << now.sec() << dendl;
+    return true;
+  }
+
+  return false;
+}
+
+std::string extract_swift_subuser(const std::string& swift_user_name) {
+  size_t pos = swift_user_name.find(':');
+  if (std::string::npos == pos) {
+    return swift_user_name;
+  } else {
+    return swift_user_name.substr(pos + 1);
+  }
+}
+
+std::string RGWTempURLAuthEngine::generate_signature(const string& key,
+                                                     const string& method,
+                                                     const string& path,
+                                                     const string& expires) const
+{
+  const string str = method + "\n" + expires + "\n" + path;
+  ldout(cct, 20) << "temp url signature (plain text): " << str << dendl;
+
+  /* unsigned */ char dest[CEPH_CRYPTO_HMACSHA1_DIGESTSIZE];
+  calc_hmac_sha1(key.c_str(), key.size(),
+                 str.c_str(), str.size(),
+                 dest);
+
+  char dest_str[CEPH_CRYPTO_HMACSHA1_DIGESTSIZE * 2 + 1];
+  buf_to_hex((const unsigned char *)dest, sizeof(dest), dest_str);
+
+  return dest_str;
+}
+
+RGWAuthApplier::aplptr_t RGWTempURLAuthEngine::authenticate() const
+{
+  const string temp_url_sig = s->info.args.get("temp_url_sig");
+  const string temp_url_expires = s->info.args.get("temp_url_expires");
+  if (temp_url_sig.empty() || temp_url_expires.empty()) {
+    return nullptr;
+  }
+
+  RGWUserInfo owner_info;
+  try {
+    get_owner_info(owner_info);
+  } catch (...) {
+    ldout(cct, 5) << "cannot get user_info of account's owner" << dendl;
+    return nullptr;
+  }
+
+  if (owner_info.temp_url_keys.empty()) {
+    ldout(cct, 5) << "user does not have temp url key set, aborting" << dendl;
+    return nullptr;
+  }
+
+  if (is_expired(temp_url_expires)) {
+    ldout(cct, 5) << "temp url link expired" << dendl;
+    return nullptr;
+  }
+
+  /* We need to verify two paths because of compliance with Swift, Tempest
+   * and old versions of RadosGW. The second item will have the prefix
+   * of Swift API entry point removed. */
+  const size_t pos = g_conf->rgw_swift_url_prefix.find_last_not_of('/') + 1;
+  const std::vector<std::string> allowed_paths = {
```

Comment:
I found a way to do this with boost::string ref (it's in a DNM PR--maybe have a look at it?)